File: Labs/FinalExamReference/fileIO/fileIO.m
% ── OCTAVE ──────────────────────────────────────────────────────────

% ── Basic Read ───────────────────────────────────────────────────────

% Read whole file as string
fid = fopen('input.txt', 'r');
if fid == -1, error('Cannot open file'); end
content = fileread('input.txt');            % easiest whole-file read
disp(content);
fclose(fid);

% Read line by line
fid = fopen('input.txt', 'r');
if fid == -1, error('Cannot open file'); end
line = fgetl(fid);                          % read first line
while ischar(line)                          % ischar = false at EOF
    disp(line);
    line = fgetl(fid);                      % read next line
end
fclose(fid);

% ── Basic Write ──────────────────────────────────────────────────────

% Write to file (overwrites)
fid = fopen('output.txt', 'w');
if fid == -1, error('Cannot open file'); end
fprintf(fid, 'Hello World\n');
fclose(fid);

% Append to file
fid = fopen('output.txt', 'a');
fprintf(fid, 'Another line\n');
fclose(fid);

% ── Read, Process, Write ─────────────────────────────────────────────
in_fid  = fopen('students.txt', 'r');
out_fid = fopen('results.txt',  'w');
if in_fid == -1,  error('Cannot open input');  end
if out_fid == -1, error('Cannot open output'); end

line = fgetl(in_fid);
while ischar(line)
    if ~isempty(strtrim(line))              % skip empty lines
        parts = strsplit(line, ',');
        name  = strtrim(parts{1});
        score = str2num(strtrim(parts{2}));
        if score >= 50
            grade = 'PASS';
        else
            grade = 'FAIL';
        end
        fprintf(out_fid, '%s | %s\n', name, grade);
    end
    line = fgetl(in_fid);
end

fclose(in_fid);
fclose(out_fid);

File: Labs/FinalExamReference/fileIO/fileIO.py
# ── PYTHON ──────────────────────────────────────────────────────────

# ── Basic Read ───────────────────────────────────────────────────────

# Read whole file as string
with open('input.txt', 'r') as f:
    content = f.read()
    print(content)

# Read as array of lines
with open('input.txt', 'r') as f:
    lines = f.readlines()                   # keeps \n at end of each line
    for line in lines:
        print(line.strip())                 # strip() removes \n

# Read line by line (memory efficient for large files)
with open('input.txt', 'r') as f:
    for line in f:
        print(line.strip())

# ── Basic Write ──────────────────────────────────────────────────────

# Write to file (overwrites)
with open('output.txt', 'w') as f:
    f.write('Hello World\n')

# Append to file
with open('output.txt', 'a') as f:
    f.write('Another line\n')

# ── Read, Process, Write ─────────────────────────────────────────────
with open('students.txt', 'r') as f:
    lines = f.readlines()

output = []
for line in lines:
    line = line.strip()
    if not line: continue                   # skip empty lines
    name, score = line.split(',')
    grade = 'PASS' if int(score) >= 50 else 'FAIL'
    output.append(f"{name.strip()} | {grade}")

with open('results.txt', 'w') as f:
    f.write('\n'.join(output))